src/openrct2/actions/SignSetStyleAction.hpp

```diff
@@ -0,0 +1,119 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include "../Context.h"
+#include "../core/MemoryStream.h"
+#include "../drawing/Drawing.h"
+#include "../localisation/StringIds.h"
+#include "../ui/UiContext.h"
+#include "../windows/Intent.h"
+#include "../world/Sprite.h"
+#include "../world/Banner.h"
+#include "GameAction.h"
+
+DEFINE_GAME_ACTION(SignSetStyleAction, GAME_COMMAND_SET_SIGN_STYLE, GameActionResult)
+{
+private:
+    int32_t _bannerIndex;
+    uint8_t _mainColour;
+    uint8_t _textColour;
+    bool _is_large;
+
+public:
+    SignSetStyleAction() = default;
+    SignSetStyleAction(int32_t bannerIndex, uint8_t mainColour, uint8_t textColour, bool is_large)
+        : _bannerIndex(bannerIndex)
+        , _mainColour(mainColour)
+        , _textColour(textColour)
+        , _is_large(is_large)
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags() | GA_FLAGS::ALLOW_WHILE_PAUSED;
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+        stream << DS_TAG(_bannerIndex) << DS_TAG(_mainColour) << DS_TAG(_textColour) << DS_TAG(_is_large);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if ((BannerIndex)_bannerIndex >= MAX_BANNERS || _bannerIndex < 0)
+        {
+            log_warning("Invalid game command for setting sign style, banner id = %d", _bannerIndex);
+            return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        return MakeResult();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_banner* banner = &gBanners[_bannerIndex];
+
+        int32_t x = banner->x << 5;
+        int32_t y = banner->y << 5;
+
+        if (_is_large)
+        {
+            TileElement* tileElement = banner_get_tile_element((BannerIndex)_bannerIndex);
+            if (tileElement == nullptr || tileElement->GetType() != TILE_ELEMENT_TYPE_LARGE_SCENERY)
```

Comment:
This part should be probably in Query as well.

Reply:
Done.

It seems like failing validations in Query need to return an error status to prevent Execute logic from running, so I've modified some of the MakeResults() accordingly. Does this look correct?